### ISSUE TITLE
rustdoc: "Namespace" user-written Markdown headers

### ISSUE THREAD

**Issue by camelid** [rust-lang/rust#91759]
Tracking issue for @jsha's idea described [here](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/DOM.20id.20tidy.20check/near/264504935).

**Comment by GuillaumeGomez** [comment]
I'll work on this.

**Comment by camelid** [comment]
@rfcbot fcp merge

Overall, I'm in favor of this idea; however, I'd still like to have a discussion about breakage before we merge in.

See https://github.com/rust-lang/rust/pull/92043#issuecomment-996995222 for my concerns.

**Comment by camelid** [comment]
@rfcbot concern breakage

I'd like to discuss [this](https://github.com/rust-lang/rust/pull/92043#issuecomment-996995222) and for us to all be on the same page before this is merged.

**Comment by camelid** [comment]
@rfcbot concern impls

I think we need to namespace by impl too.

For example, the `primitive.pointer.html` page has impls for both `*const T` and `*mut T`. The methods on each are currently disambiguated using the number suffixes (cc @scottmcm who ran into issues with this), but after this change, I think they would conflict. IIUC, the current implementation in #92043 only namespaces by the parent item (e.g., method name), but what about when there are two impls on a page with the same methods names?

Note that this occurs for some user types too, not just the pointer page.

**Comment by jsha** [comment]
Good point. Here's a concrete example:

https://doc.rust-lang.org/nightly/std/any/trait.Any.html#examples-7
https://doc.rust-lang.org/nightly/std/any/trait.Any.html#examples-2

Those are the examples section for `<dyn Any + 'static>::downcast_ref` and `<dyn Any + Send + 'static>::downcast_ref`, respectively.

The links to those impls themselves have the same "incrementing number" problem: https://doc.rust-lang.org/nightly/std/any/trait.Any.html#impl and https://doc.rust-lang.org/nightly/std/any/trait.Any.html#impl-1. So it would be good to fix that at the same time we fix this.

For instance, we might make the link to the first impl `#impl.dyn-any-static`, and the second `#impl.dyn-any-send-static`. Then the examples sections could be `#impl.dyn-any-static.method.downcast_ref.examples` or similar.

Note: there is prior art for certain impls getting an anchor that reflects their type parameters. For instance see https://doc.rust-lang.org/nightly/std/string/struct.String.html#impl-Add%3C%26%27_%20str%3E.

This happens in the "Implementations" section of a struct page. It doesn't happen in the "Implementors" section of a trait page.

Note that the String anchor linked above is [over-encoded](https://url.spec.whatwg.org/#fragment-percent-encode-set). If we wind up tweaking generation of impl targets I'd like something with the right level of escaping. [See this demo](https://rustdoc.crud.net/jsha/anchor.html#impl-Add<&'_+str>).

    <a id="impl-Add<&'_ str>" href="#impl-Add%3C&'_ str%3E">target</a>

**Comment by camelid** [comment]
I agree with what you say, but I mean that headings _within_ methods _within_ impls need to be namespaced like `#impl-Add.method.add.examples`.

**Comment by jsha** [comment]
Yep, we're on the same page - I was intending to agree and expand. In order to fix the conflict between same-named methods, we need to namespace by impl plus method name; that in turn brings up the parallel issue that id assignment for impls has a very similar problem, and we should fix both at once.

**Comment by GuillaumeGomez** [comment]
Both of what you suggested (keeping the whole "path" and changing the impls ID) will greatly increase the size of the generated IDs and therefore the size of the generated output. I don't think it's worth it to keep the whole path, however changing the impls ID sounds like a good idea.

**Comment by camelid** [comment]
We need to keep the whole path or we'll introduce ID conflicts.

**Comment by GuillaumeGomez** [comment]
No, it's handled by the `IdMap` just like currently.

**Comment by camelid** [comment]
But then we have the same issue of the IDs not being stable. I think we should make our IDs simpler and "more correct" by ensuring they are truly unanmbiguous (or as little ambiguity as possible). Compression will probably mitigate the page size cost, right?

**Comment by GuillaumeGomez** [comment]
As I understood, the main goal was to not have conflicts with rustdoc's IDs. With the current implementation it's already quite limited.

**Comment by jsha** [comment]
> Both of what you suggested (keeping the whole "path" and changing the impls ID) will greatly increase the size of the generated IDs and therefore the size of the generated output.

One possible size optimization: for types with a single `impl` block, omit the impl prefix and just do `#method.foo.examples`.

By the way, a minor problem with my proposal above about generating better ids for impl blocks: a type can have multiple impl blocks with the same type parameters (or no type parameters), so we would still need some form of collision-renaming like we have today.

**Comment by camelid** [comment]
> One possible size optimization: for types with a single `impl` block, omit the impl prefix and just do `#method.foo.examples`.

Then adding an impl will still be cause link breakage. The page size savings don't seem worth it to me. IMO it'd be better for rustdoc's rules to be simple and correct.

> By the way, a minor problem with my proposal above about generating better ids for impl blocks: a type can have multiple impl blocks with the same type parameters (or no type parameters), so we would still need some form of collision-renaming like we have today.

Note that that's only a problem for the impl headings themselves since the methods within must have unique names or rustc will get upset about naming conflicts.

-----

I think this has gotten a bit off-topic, sorry about that! The goal of **this** proposal is to **prevent conflicts and reduce instability for user-written Markdown headings**.

I think we should start **another** proposal to **reduce instability of rustdoc-generated headings (impls and associated items)**.

**Comment by camelid** [comment]
Ok, I've opened #92052. I haven't started FCP there yet because I think we should finish with this change first to avoid too discussing too many changes at once.

@rfcbot resolve impls

I think the current approach should probably be good enough. I think it should be overall an improvement.

**Comment by GuillaumeGomez** [comment]
@rfcbot approve

**Comment by camelid** [comment]
@Nemo157 could you review this when you get a chance? If you have any concerns, I'd be happy to discuss them here.

**Comment by camelid** [comment]
Thanks Nemo! The only thing blocking this now is my "breakage" concern. Is everyone okay with the breakage this change will cause?

@rfcbot concern anchor formatting

The one other concern I might propose is the choice of formatting for the anchor prefixes. Should they (roughly) match their parent, e.g., `#method.foo.my-anchor-here`? Or should they be just based on namespaces, e.g., `#v.foo.my-anchor-here`? Or something else? We could always change this later, perhaps with a JS-powered anchor redirect for old anchors, but it'd be best to avoid doing that if possible.

**Comment by GuillaumeGomez** [comment]
What is the `v` in `v.foo...`? But otherwise, either way is fine to me. The second has the advantage to be smaller though.

**Comment by Manishearth** [comment]
`value`, since rust has `value`, `type`, and `macro` namespaces. But the namespaces don't matter here, for associated stuff it's all the same namespace, I _think_.

**Comment by jsha** [comment]
> Should they (roughly) match their parent, e.g., `#method.foo.my-anchor-here`

This is my preferred approach.

**Comment by GuillaumeGomez** [comment]
Let's go for the expanded name then? It's currently how it's implemented in any case so less work for me. :3

**Comment by scottmcm** [comment]
> Edit: nope, you can class associated items.

Yeah, there's a couple weird cases here -- see https://github.com/rust-lang/rust/issues/76347 for example.

**Comment by camelid** [comment]
@rfcbot resolve anchor formatting

Seems good for now, we can probably figure out a way to change it later if necessary.

**Comment by camelid** [comment]
@rfcbot resolve breakage

No one's raised concerns about the breakage (this concern was just a placeholder), so I'll resolve this concern. The links broke very easily before anyway.

**Comment by jsha** [comment]
@rfcbot concern simplify

I'd like to propose a slightly simpler alternative: We could not to generate anchors for most user-generated markdown.

The headings in the top-doc are usually substantive, and worth linking to on their own, so we could namespace those. For instance https://doc.rust-lang.org/nightly/std/string/struct.String.html#utf-8 would become https://doc.rust-lang.org/nightly/std/string/struct.String.html#top-doc.utf-8.

But for other places, like https://doc.rust-lang.org/nightly/std/string/struct.String.html#examples-23, users could just as well link to the parent item, https://doc.rust-lang.org/nightly/std/string/struct.String.html#method.as_bytes. For most methods and other items outside the top-doc, having links to specific user sub-headings may not be worth the complexity we add by trying to namespace all of them.

Rustdoc currently never generates internal links to user sub-headings, so these anchors are purely an external-facing interface. Right now that interface is mostly broken - the numbered anchors are very subject to breakage. Rather than fixing that interface, maybe we should narrow it.

**Comment by camelid** [comment]
Hmm, seems interesting. I don't feel super strongly either way. What do other people think?

**Comment by GuillaumeGomez** [comment]
This is an interesting take and would simplify rustdoc code a lot. I think it's a good idea so I'm favour for it but not feeling strongly about it either.

**Comment by camelid** [comment]
If we do end up doing this, I think we should still keep anchors for headings in impl docs. Just a couple days ago I used those to split up some docs into sections.

**Comment by lolbinarycat** [comment]
Regarding breakage, would it not be possible to have a crate-level attribute that controls the fragment behavior, then change the default value over an edition?

I've also seen conflicts due to a user-written heading having the same name as an auto-generated heading (eg. due to someone writing `# Functions` in their crate-level docs, which I believe can also mess up the sidebar?)

It shouldn't be too much work to keep the legacy behavior around since it basically just boils down to "skip adding prefixes before disambiguating with numbers".  The numbers code would need to stick around anyways in case someone writes docs like this:
```rs
/// # Foo
/// ...
/// ## Examples
/// ...
/// # Bar
/// ...
/// ## Examples
/// ...
```
